actionlib/simple_action_server.go

```diff
@@ -0,0 +1,295 @@
+
+func (s *simpleActionServer) Start() {
+	if s.executeCb != nil {
+		go s.goalExecutor()
+	}
+	s.actionServer.Start()
+}
+
+func (s *simpleActionServer) internalGoalCallback(goal ActionGoal) {
+	s.logger.Infof("Simple action server received new goal with id %s", goal.GetGoalId().Id)
+
+	goalStamp := goal.GetGoalId().Stamp
+	nextGoalStamp := goal.GetGoalId().Stamp
+	newGoal := newStatusWithActionGoal(s.actionServer, goal)
+
+	s.goalMutex.Lock()
+	defer s.goalMutex.Unlock()
+
+	if (s.currentGoal == nil || goalStamp.Cmp(s.currentGoal.getGoalID().Stamp) >= 0) &&
+		(s.currentGoal == nil || nextGoalStamp.Cmp(s.currentGoal.getGoalID().Stamp) >= 0) {
+
+		if s.nextGoal != nil && (s.currentGoal == nil || s.nextGoal != s.currentGoal) {
+			s.nextGoal.setCancelled(nil,
+				"This goal was canceled because another goal was received by the simple action server")
+		}
+
+		s.nextGoal = newGoal
+		s.newGoal = true
+		s.newGoalPreemptRequest = false
+		args := []reflect.Value{reflect.ValueOf(goal)}
+
+		if s.IsActive() {
+			s.preemptRequest = true
+			if err := s.runCallback("preempt", args); err != nil {
+				s.logger.Error(err)
+			}
+		}
+
+		if err := s.runCallback("goal", args); err != nil {
+			s.logger.Error(err)
+		}
+
+		// notify executor that a new goal is available
+		s.executorCh <- struct{}{}
+	} else {
+		newGoal.setCancelled(nil,
+			"This goal was canceled because another goal was received by the simple action server")
+	}
+}
+
+func (s *simpleActionServer) internalPreemptCallback(preempt ActionGoal) {
+	s.goalMutex.Lock()
+	defer s.goalMutex.Unlock()
+
+	s.logger.Infof("Simple action server received preempt call for goal with id %s",
+		preempt.GetGoalId().Id)
+
+	if preempt == s.currentGoal.getGoal() {
+		s.preemptRequest = true
+		args := []reflect.Value{reflect.ValueOf(preempt)}
+		if err := s.runCallback("preempt", args); err != nil {
+			s.logger.Error(err)
+		}
+	} else {
+		s.newGoalPreemptRequest = true
+	}
+}
+
+func (s *simpleActionServer) IsNewGoalAvailable() bool {
+	s.goalMutex.Lock()
+	defer s.goalMutex.Unlock()
+
+	return s.newGoal
+}
+
+func (s *simpleActionServer) IsPreemptRequested() bool {
+	s.goalMutex.Lock()
+	defer s.goalMutex.Unlock()
+
+	return s.preemptRequest
+}
+
+func (s *simpleActionServer) AcceptNewGoal() (ActionGoal, error) {
+	s.goalMutex.Lock()
+	defer s.goalMutex.Unlock()
+
+	if !s.newGoal || s.nextGoal.getGoal() == nil {
+		return nil, fmt.Errorf("Attempting to accept the next goal when a new goal is not available")
+	}
+
+	if s.IsActive() && s.currentGoal.getGoal() != nil && s.currentGoal != s.nextGoal {
+		s.currentGoal.setCancelled(nil,
+			"This goal was canceled because another goal was received by the simple action server")
+	}
+
+	fmt.Println("Accepting new goal")
+	s.currentGoal = s.nextGoal
+	s.newGoal = false
+	s.preemptRequest = false
+	s.newGoalPreemptRequest = false
+
+	s.currentGoal.setAccepted("This goal has been accepted by the simple action server")
+
+	return s.currentGoal.getGoal(), nil
```

Comment:
This is a dumb question but given the getGoal is the return call, I suppose that Golang is smart enough to not let go of the mutex until that call is calculated... or are defer statements run before the return

Reply:
Defer statements are run after the return. So, this method should still have the mutex lock while calling getGoal.